src/main/java/org/springframework/data/repository/query/QueryByExampleExecutor.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.repository.query;
+
+import org.springframework.data.domain.Example;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Sort;
+
+/**
+ * Interface to allow execution of Query by Example {@link Example} instances.
+ *
+ * @author Mark Paluch
+ * @since 1.12
+ */
+public interface QueryByExampleExecutor<T> {
+
+	/**
+	 * Returns a single entity matching the given {@link Example} or {@literal null} if none was found.
+	 *
+	 * @param Example can be {@literal null}.
+	 * @return a single entity matching the given {@link Example} or {@literal null} if none was found.
+	 * @throws org.springframework.dao.IncorrectResultSizeDataAccessException if the Example yields more than one result.
+	 */
+	<S extends T> T findOne(Example<S> example);
```

Comment:
Had a second look at the declaration. In fact, a Repository will always return `T` because the type is looked up from the repository, not the probe. I'd change the declaration to `T findOne(Example<? extends T> example)`, same for the other methods.